File: Day12/Part2/portal_bfs.py
"""
Day 12 — Part 2: Portal Teleportation (BFS with State)

Add portal mechanics:
- 'P' cells allow one-time teleportation to any other 'P'
- Use BFS with state: (row, col, usedPortal)

Approach: BFS with extended state
Time: O(n*m * 2), Space: O(n*m * 2)
"""
from typing import List
from collections import deque


def bfs_with_portal(grid: List[str]) -> int:
    """
    Find shortest path with one-time portal teleportation.
    Returns minimum steps or -1 if unreachable.
    """
    n, m = len(grid), len(grid[0])
    
    # Find start, end, and all portals
    start = end = None
    portals = []
    
    for i in range(n):
        for j in range(m):
            if grid[i][j] == 'S':
                start = (i, j)
            elif grid[i][j] == 'E':
                end = (i, j)
            elif grid[i][j] == 'P':
                portals.append((i, j))
    
    if not start or not end:
        return -1
    
    # BFS with state: (row, col, usedPortal, steps)
    q = deque([(start[0], start[1], False, 0)])
    visited = set()
    visited.add((start[0], start[1], False))
    
    directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
    
    while q:
        r, c, used_portal, steps = q.popleft()
        
        # Check if reached end
        if (r, c) == end:
            return steps
        
        # Normal movement
        for dr, dc in directions:
            nr, nc = r + dr, c + dc
            
            if 0 <= nr < n and 0 <= nc < m and (nr, nc, used_portal) not in visited:
                if grid[nr][nc] != '#':
                    visited.add((nr, nc, used_portal))
                    q.append((nr, nc, used_portal, steps + 1))
        
        # Portal teleportation (if standing on portal and haven't used it)
        if grid[r][c] == 'P' and not used_portal:
            for pr, pc in portals:
                if (pr, pc) != (r, c) and (pr, pc, True) not in visited:
                    visited.add((pr, pc, True))
                    q.append((pr, pc, True, steps + 1))
    
    return -1


if __name__ == "__main__":
    # Example from problem
    grid1 = [
        "S.#.",
        ".P..",
        "..#E",
        "P..."
    ]
    result1 = bfs_with_portal(grid1)
    print(f"Part2 Example: {result1} steps (with portal)")
    assert result1 > 0, f"Should find a path, got {result1}"
    
    # Test with beneficial portal use
    grid2 = [
        "S.....",
        "......",
        "P....P",
        "......",
        "......E"
    ]
    result2 = bfs_with_portal(grid2)
    print(f"Part2 Test2: {result2} steps")
    
    print("Day12 Part2 portal_bfs.py: tests passed")

File: learnLeetCode/Day12/Part2/portal_bfs.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <set>
#include <string>
#include <tuple>
using namespace std;

int bfsWithPortal(const vector<string>& grid) {
    int n = grid.size();
    if (n == 0) return -1;
    int m = grid[0].size();
    
    // Find start, end, and portals
    pair<int,int> start = {-1, -1}, end = {-1, -1};
    vector<pair<int,int>> portals;
    
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            if (grid[i][j] == 'S') start = {i, j};
            else if (grid[i][j] == 'E') end = {i, j};
            else if (grid[i][j] == 'P') portals.push_back({i, j});
        }
    }
    
    if (start.first == -1 || end.first == -1) return -1;
    
    // BFS with state: (row, col, usedPortal, steps)
    queue<tuple<int,int,bool,int>> q;
    q.push({start.first, start.second, false, 0});
    set<tuple<int,int,bool>> visited;
    visited.insert({start.first, start.second, false});
    
    int dr[] = {0, 0, 1, -1};
    int dc[] = {1, -1, 0, 0};
    
    while (!q.empty()) {
        auto [r, c, usedPortal, steps] = q.front();
        q.pop();
        
        if (r == end.first && c == end.second) return steps;
        
        // Normal movement
        for (int i = 0; i < 4; ++i) {
            int nr = r + dr[i], nc = c + dc[i];
            
            if (nr >= 0 && nr < n && nc >= 0 && nc < m && 
                visited.find({nr, nc, usedPortal}) == visited.end()) {
                if (grid[nr][nc] != '#') {
                    visited.insert({nr, nc, usedPortal});
                    q.push({nr, nc, usedPortal, steps + 1});
                }
            }
        }
        
        // Portal teleportation
        if (grid[r][c] == 'P' && !usedPortal) {
            for (auto [pr, pc] : portals) {
                if ((pr != r || pc != c) && visited.find({pr, pc, true}) == visited.end()) {
                    visited.insert({pr, pc, true});
                    q.push({pr, pc, true, steps + 1});
                }
            }
        }
    }
    
    return -1;
}

int main() {
    // Example
    vector<string> grid1 = {
        "S.#.",
        ".P..",
        "..#E",
        "P..."
    };
    int result1 = bfsWithPortal(grid1);
    cout << "Part2 Example: " << result1 << " steps (with portal)" << endl;
    if (result1 <= 0) return 1;
    
    // Test with portals
    vector<string> grid2 = {
        "S.....",
        "......",
        "P....P",
        "......",
        "......E"
    };
    int result2 = bfsWithPortal(grid2);
    cout << "Part2 Test2: " << result2 << " steps" << endl;
    
    cout << "Day12 Part2 portal_bfs.cpp: tests passed" << endl;
    return 0;
}
